fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushDownNotSlotReferenceExpressionOfOnClause.java

```diff
@@ -0,0 +1,200 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * push down expression which is not slot reference
+ */
+public class PushDownNotSlotReferenceExpressionOfOnClause extends OneRewriteRuleFactory {
+
+    private static class ListBuilder<T> {
+        List<T> list;
+
+        public ListBuilder(List<T> list) {
+            this.list = Lists.newArrayList(list);
+        }
+
+        public ListBuilder<T> addAll(List<T> element) {
+            list.addAll(Lists.newArrayList(element));
+            return this;
+        }
+
+        public List<T> get() {
+            return list;
+        }
+    }
+
+    private final Map<SlotReference, Map<Expression, Expression>> exprMap = new HashMap<>();
+
+    /**
+     * rewrite example:
+     *join(t1.a + 1 = t2.b + 2 and t1.a > 2)        join(c = d and c > 2)
+     *          │                                             │
+     *          ├─olapScan(t1)                                ├────project(t1.a + 1 as c)
+     *          │                               ====>         │       │
+     *          └─olapScan(t2)                                │       └─olapScan(t1)
+     *                                                        │
+     *                                                        └────project(t2.b + 2 as d)
+     *                                                                │
+     *                                                                └─olapScan(t2)
+     */
+    @Override
+    public Rule build() {
+        return logicalJoin()
+                .when(join -> join.getCondition().isPresent()
+                        && ((ImmutableList) join.getCondition().get()
+                        .collect(Arithmetic.class::isInstance)).size() > 0)
+                .then(join -> {
+                    join.getCondition().get().accept(
+                            new TreeCollectVisitor(),
+                            exprMap);
+
+                    List<Plan> childs = join.children().stream()
+                            .map(GroupPlan.class::cast)
+                            .map(p -> {
+                                List<NamedExpression> slots = p.getOutput()
+                                                .stream()
+                                                .filter(SlotReference.class::isInstance)
+                                                .filter(ref -> exprMap.containsKey(ref))
+                                                .map(ref -> exprMap.get(ref).values())
+                                                .flatMap(Collection::stream)
+                                                .map(NamedExpression.class::cast)
+                                                .collect(Collectors.toList());
+                                return new LogicalProject<>(slots, p.getGroup().getLogicalExpression().getPlan());
```

Comment:
u need to use join's children as ur children, and cannot get plan from memo in rules